python/lsst/daf/butler/direct_query_driver/_predicate_constraints_summary.py

```diff
@@ -0,0 +1,221 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This software is dual licensed under the GNU General Public License and also
+# under a 3-clause BSD license. Recipients may choose which of these licenses
+# to use; please see the files gpl-3.0.txt and/or bsd_license.txt,
+# respectively.  If you choose the GPL option then the following text applies
+# (but note that there is still no warranty even if you opt for BSD instead):
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+from typing import Any
+
+from .._exceptions import InvalidQueryError
+from ..dimensions import DataCoordinate, DataIdValue, DimensionGroup, DimensionUniverse
+from ..queries import tree as qt
+from ..queries.visitors import ColumnExpressionVisitor, PredicateVisitFlags, SimplePredicateVisitor
+
+
+class PredicateConstraintsSummary:
+    """Summarizes information about the constraints on data ID values implied
+    by a Predicate.
+
+    Parameters
+    ----------
+    predicate : `Predicate`
+        Predicate to summarize.
+    """
+
+    predicate: qt.Predicate
+    """The predicate examined by this summary."""
+
+    constraint_data_id: dict[str, DataIdValue]
+    """Data ID values that will be identical in all result rows due to query
+    constraints.
+    """
+
+    messages: list[str]
+    """Diagnostic messages that report reasons the query may not return any
+    rows.
+    """
+
+    _governors_referenced: set[str]
+    """Governor dimensions referenced directly in the predicate, but not
+    necessarily constrained to the same value in all logic branches.
+    """
```

Comment:
I find it a little strange stylistically to annotate and document a private attribute out here, rather than just annotating it in `__init__`.